### ISSUE TITLE
bear -- make failed when ipv6 was not supported

### ISSUE THREAD

**Issue by changvy** [rizsotto/Bear#296]
D0922 11:42:02.883185411   16499 is_epollexclusive_available.cc:86] epoll_ctl with EPOLLEXCLUSIVE | EPOLLONESHOT succeeded. This is evidence of no EPOLLEXCLUSIVE support. Not using epollex polling engine.
I0922 11:42:02.883308582   16499 ev_epollex_linux.cc:1624]   Skipping epollex because it is not supported.
I0922 11:42:02.883389185   16499 ev_epoll1_linux.cc:122]     grpc epoll fd: 3
D0922 11:42:02.883406489   16499 ev_posix.cc:173]            Using polling engine: epoll1
D0922 11:42:02.883501703   16499 lb_policy_registry.cc:40]   registering LB policy factory for "grpclb"
D0922 11:42:02.883524650   16499 lb_policy_registry.cc:40]   registering LB policy factory for "cds_experimental"
D0922 11:42:02.883540746   16499 lb_policy_registry.cc:40]   registering LB policy factory for "eds_experimental"
D0922 11:42:02.883560209   16499 lb_policy_registry.cc:40]   registering LB policy factory for "lrs_experimental"
D0922 11:42:02.883580485   16499 lb_policy_registry.cc:40]   registering LB policy factory for "priority_experimental"
D0922 11:42:02.883601656   16499 lb_policy_registry.cc:40]   registering LB policy factory for "weighted_target_experimental"
D0922 11:42:02.883623747   16499 lb_policy_registry.cc:40]   registering LB policy factory for "xds_routing_experimental"
D0922 11:42:02.883641182   16499 lb_policy_registry.cc:40]   registering LB policy factory for "pick_first"
D0922 11:42:02.883661799   16499 lb_policy_registry.cc:40]   registering LB policy factory for "round_robin"
D0922 11:42:02.883681506   16499 dns_resolver_ares.cc:507]   Using ares dns resolver
I0922 11:42:02.883783171   16499 server_builder.cc:319]      Synchronous server. Num CQs: 1, Min pollers: 1, Max Pollers: 2, CQ timeout (msec): 10000
I0922 11:42:02.883999944   16499 socket_utils_common_posix.cc:427] Disabling AF_INET6 sockets because ::1 is not available.
I0922 11:42:02.884048028   16499 socket_utils_common_posix.cc:352] TCP_USER_TIMEOUT is available. TCP_USER_TIMEOUT will be used thereafter
I0922 11:42:02.884153329   16499 tcp_server_posix.cc:325]    Failed to add :: listener, the environment may not support IPv6: {"created":"@1600746122.884039943","description":"Address family not supported by protocol","errno":97,"file":"/data/soft/bear/Bear/subprojects/Source/grpc_dependency/src/core/lib/iomgr/socket_utils_common_posix.cc","file_line":443,"os_error":"Address family not supported by protocol","syscall":"socket","target_address":"[::]:0"}
D0922 11:42:02.887029084   16504 is_epollexclusive_available.cc:86] epoll_ctl with EPOLLEXCLUSIVE | EPOLLONESHOT succeeded. This is evidence of no EPOLLEXCLUSIVE support. Not using epollex polling engine.
I0922 11:42:02.887139801   16504 ev_epollex_linux.cc:1624]   Skipping epollex because it is not supported.
I0922 11:42:02.887158196   16504 ev_epoll1_linux.cc:122]     grpc epoll fd: 3
D0922 11:42:02.887178745   16504 ev_posix.cc:173]            Using polling engine: epoll1
D0922 11:42:02.887298022   16504 lb_policy_registry.cc:40]   registering LB policy factory for "grpclb"
D0922 11:42:02.887324727   16504 lb_policy_registry.cc:40]   registering LB policy factory for "cds_experimental"
D0922 11:42:02.887343801   16504 lb_policy_registry.cc:40]   registering LB policy factory for "eds_experimental"
D0922 11:42:02.887363067   16504 lb_policy_registry.cc:40]   registering LB policy factory for "lrs_experimental"
D0922 11:42:02.887384467   16504 lb_policy_registry.cc:40]   registering LB policy factory for "priority_experimental"
D0922 11:42:02.887406174   16504 lb_policy_registry.cc:40]   registering LB policy factory for "weighted_target_experimental"
D0922 11:42:02.887429556   16504 lb_policy_registry.cc:40]   registering LB policy factory for "xds_routing_experimental"
D0922 11:42:02.887450006   16504 lb_policy_registry.cc:40]   registering LB policy factory for "pick_first"
D0922 11:42:02.887470769   16504 lb_policy_registry.cc:40]   registering LB policy factory for "round_robin"
D0922 11:42:02.887490100   16504 dns_resolver_ares.cc:507]   Using ares dns resolver
I0922 11:42:02.888067836   16504 socket_utils_common_posix.cc:427] Disabling AF_INET6 sockets because ::1 is not available.
D0922 11:42:02.899157134   16504 tcp_posix.cc:1830]          cannot set inq fd=5 errno=92
I0922 11:42:02.899331539   16504 http_connect_handshaker.cc:329] Connecting to server 0.0.0.0:38352 via HTTP proxy ipv4:10.14.36.121:8080
I0922 11:42:02.904573807   16504 subchannel.cc:1073]         Connect failed: {"created":"@1600746122.904521385","description":"HTTP proxy returned response code 502","file":"/data/soft/bear/Bear/subprojects/Source/grpc_dependency/src/core/ext/filters/client_channel/http_connect_handshaker.cc","file_line":246}
I0922 11:42:02.904651817   16504 subchannel.cc:1012]         Subchannel 0x1125200: Retry in 990 milliseconds
er: failed with: gRPC call failed: failed to connect to all addresses [pid: 16504]

**Comment by rizsotto** [comment]
Can you tell me more about your set up? What OS do you use?

**Comment by changvy** [comment]
Linux VM_54_114_centos 3.10.107-1-tlinux2_kvm_guest-0049 #1 SMP Tue Jul 30 23:46:29 CST 2019 x86_64 x86_64 x86_64 GNU/Linux
gcc version 9.3.1 20200408 (Red Hat 9.3.1-2) (GCC)

This is the version of the operation system, my colleague made some changes based on centos as you see.  
Gcc version is 9.3.1 with the help of scl.

**Comment by rizsotto** [comment]
Thanks @changvy to come back.

I failed to reproduce it on a machine which I've disabled the IPv6 addresses. Can you help me to test this on your machine? Is this a fix on your machine? (If you save this into a file `/tmp/bear.diff` then just `git apply /tmp/bear.diff` will make the change.)

```diff
diff --git a/source/intercept/source/Application.cc b/source/intercept/source/Application.cc
index d36b9254..d85cfdc7 100644
--- a/source/intercept/source/Application.cc
+++ b/source/intercept/source/Application.cc
@@ -106,7 +106,7 @@ namespace ic {
                           .AddListeningPort("0.0.0.0:0", grpc::InsecureServerCredentials(), &port)
                           .BuildAndStart();
 
-        std::string server_address = fmt::format("0.0.0.0:{}", port);
+        std::string server_address = fmt::format("127.0.0.1:{}", port);
         spdlog::debug("Running gRPC server. [Listening on {0}]", server_address);
         // Configure the session and the reporter objects
         impl_->session_->set_server_address(server_address);
```

**Comment by Cnly** [comment]
I also came across this problem recently. For me the solution was to remove the HTTP proxy environment variables I exported.

**Comment by rizsotto** [comment]
Updated the code to use `12.7.0.0.1` address. Which shall work with HTTP proxies (presume you have `no_proxy=localhost,127.0.0.1` also set).

Fix will be in 3.0.4 release.

**Comment by ensc** [comment]
What happens when IPv4 is disabled and `127.0.0.1` can not be used?  Would it be better to use `localhost` instead of an IP address?
